### PR TITLE
Fix BoundedInt Add libfunc

### LINKED ISSUE #1448: Review bounded_int_add libfunc
Review the `bounded_int_add` libfunc, to ensure that all possible inputs are taken into account.

- What types can the libfunc arguments have?
- Is the offset of all bounded int values taken into account?
- Is the value of the return bounded int (if any) computed correctly, according to its offset?

We should:

- Document the signature the signature of the libfunc, including input and output values.
- Document the expected behavior of the libfunc.
- Add  unit/integration tests for edge cases.

Currently, the libfunc fails with the following arguments. When `A: BoundedInt<1,31> = 31` and `B: BoundedInt<-1,-1> = -1`, the expected return value is 30, but it actually returns 31.

Also, If the bounded int overflows, is is suppose to wrap like a felt does?

### PR DESCRIPTION
**Pull request by DiegoCivi**
# Fix BoundedInt Add libfunc

The previous implementation of the add libfunc had some cases in which it failed to return the correct result. We decided to change the algorithm to one that is simpler to understand and that fixes the errors.

We have C = A + B, where all of them are `BoundedInt`. We represent the a `BoundedInt` as: 
- X = Xo + Xd, where 
    - X is a `BoundedInt`.
    - Xo is the lower bound of the `BoundedInt` range.
    - Xd is the offset.

So we want to get `Cd` and we have `Co + Cd = Ao + Ad + Bo + Bd` -> `Cd = Ao + Ad + Bo + Bd - Co`. When having `BoundedInt` additions we need to implement the `AddHelper` like the following:
```cairo
impl AddHelper of AddHelper<BoundedInt<1, 30>, BoundedInt<0, 10>> {
    type Result = BoundedInt<1, 40>;
}

fn run_addition() -> BoundedInt<1, 40> {
    let a: BoundedInt<1, 30> = 10.try_into().unwrap();
    let b: BoundedInt<0, 10> = 5.try_into().unwrap();
    return add(a, b);
}
```
In this case we have:
- Ao = 1
- Ad = 9
- Bo = 0
- Bd = 5
- Co = 1

And as we can see `Co = Ao + Bo`. If this is not satisfied, then the Cairo code won´t compile. Then, by having `Cd = Ao + Ad + Bo + Bd - Co` and `Co = Ao + Bo` we can simplify the addition to `Cd = Ad + Bd`.

Closes #1448 

## Introduces Breaking Changes?

No.

## Checklist

- [x] Linked to Github Issue.
- [x] Unit tests added.
- [ ] Integration tests added.
- [ ] This change requires new documentation.
  - [ ] Documentation has been added/updated.

### CODE REVIEW

**Review by JulianGCalderon — CHANGES_REQUESTED** [review]

File: src/libfuncs/bounded_int.rs [hunk 1]
```diff
@@ -73,6 +73,11 @@ pub fn build<'ctx, 'this>(
 }
 
 /// Generate MLIR operations for the `bounded_int_add` libfunc.
+///
+/// Since we want to get C = A + B, we can translate this to
+/// Co + Cd = Ao + Ad + Bo + Bd. Where Ao, Bo and Co represent the lower bound
+/// of the ranges in the BoundedInt and Ad, Bd and Cd represent the offsets. Since
+/// we also know that Co = Ao + Bo we can simplify the equation to Cd = Ad + Bd.
```
Comment (JulianGCalderon) [inline src/libfuncs/bounded_int.rs]: This is a bit confusing without explaining first the signature of the libfunc, and how it is internally represented. Can you add some documentation on that?

File: src/libfuncs/bounded_int.rs [hunk 2]
```diff
@@ -1007,6 +982,140 @@ mod test {
...
+    #[test_case("run_test_5", -5, -5, -10, -10, -9)]
+    #[test_case("run_test_6", -5, -1, -6, -6, -5)]
+    #[test_case("run_test_7", 1, -5, -4, -4, -3)]
+    fn test_add(
+        entry_point: &str,
+        lhs: i32,
+        rhs: i32,
+        result: i32,
+        lower_bound: i32,
+        upper_bound: i32,
```
[the remark this answers is not in the document]
Comment (DiegoCivi) [reply src/libfuncs/bounded_int.rs]: It wasn't much trouble. Done in [`d3bae9b`](https://github.com/lambdaclass/cairo_native/pull/1463/commits/d3bae9b9938daa3c7b0e5438fa0f870b1ff40725)

File: src/libfuncs/bounded_int.rs (hunk 1 above)
Comment (DiegoCivi) [reply src/libfuncs/bounded_int.rs]: Done in [`e7da658`](https://github.com/lambdaclass/cairo_native/pull/1463/commits/e7da658866a7aecf321a4a3355d42e1ce07bc078)

File: src/libfuncs/bounded_int.rs [hunk 3]
```diff
@@ -106,31 +112,23 @@ fn build_add<'ctx, 'this>(
...
     native_assert!(
-        compute_range.offset_bit_width() >= lhs_width,
+        compute_width >= lhs_width,
         "the lhs_range bit_width must be less or equal than the compute_range"
     );
     native_assert!(
-        compute_range.offset_bit_width() >= rhs_width,
+        compute_width >= rhs_width,
         "the rhs_range bit_width must be less or equal than the compute_range"
     );
```
[the remark this answers is not in the document]
Comment (DiegoCivi) [reply src/libfuncs/bounded_int.rs]: Done in [`ec7491d`](https://github.com/lambdaclass/cairo_native/pull/1463/commits/ec7491dffd118a9f96c0c3309df525344e266e8e)

**Review by gabrielbosio — APPROVED** [review]
